salt/modules/config.py

```diff
@@ -130,31 +136,110 @@ def valid_fileproto(uri):
 
 def option(
         value,
-        default='',
+        default=None,
         omit_opts=False,
+        omit_grains=False,
+        omit_pillar=False,
         omit_master=False,
-        omit_pillar=False):
+        omit_all=False,
+        wildcard=False):
     '''
-    Pass in a generic option and receive the value that will be assigned
+    Returns the setting for the specified config value. The priority for
+    matches is the same as in :py:func:`config.get <salt.modules.config.get>`,
+    only this function does not recurse into nested data structures. Another
+    difference between this function and :py:func:`config.get
+    <salt.modules.config.get>` is that it comes with a set of "sane defaults".
+    To view these, you can run the following command:
+
+    .. code-block:: bash
+
+        salt '*' config.option '*' omit_all=True wildcard=True
+
+    default
+        The default value if no match is found. If not specified, then the
+        fallback default will be an empty string, unless ``wildcard=True``, in
+        which case the return will be an empty dictionary.
+
+    omit_opts : False
+        Pass as ``True`` to exclude matches from the minion configuration file
+
+    omit_grains : False
+        Pass as ``True`` to exclude matches from the grains
+
+    omit_pillar : False
+        Pass as ``True`` to exclude matches from the pillar data
+
+    omit_master : False
+        Pass as ``True`` to exclude matches from the master configuration file
+
+    omit_all : True
+        Shorthand to omit all of the above and return matches only from the
+        "sane defaults".
+
+        .. versionadded:: Neon
+
+    wildcard : False
+        If used, this will perform pattern matching on keys. Note that this
+        will also significantly change the return data. Instead of only a value
+        being returned, a dictionary mapping the matched keys to their values
+        is returned. For example, using ``wildcard=True`` with a ``key`` of
+        ``'foo.ba*`` could return a dictionary like so:
+
+        .. code-block:: python
+
+            {'foo.bar': True, 'foo.baz': False}
+
+        .. versionadded:: Neon
 
     CLI Example:
 
     .. code-block:: bash
 
         salt '*' config.option redis.host
     '''
-    if not omit_opts:
-        if value in __opts__:
-            return __opts__[value]
-    if not omit_master:
-        if value in __pillar__.get('master', {}):
-            return __pillar__['master'][value]
-    if not omit_pillar:
-        if value in __pillar__:
-            return __pillar__[value]
-    if value in DEFAULTS:
-        return DEFAULTS[value]
-    return default
+    if omit_all:
+        omit_opts = omit_grains = omit_pillar = omit_master = True
+
+    if default is None:
+        default = '' if not wildcard else {}
+
+    if not wildcard:
+        if not omit_opts:
+            if value in __opts__:
+                return __opts__[value]
```

Comment:
Would it be better to do

```
try:
    return __opts__[value]
except KeyError:
    pass
```
? (and then the same things for the later?)

Actually, if I understand this block correctly, couldn't you write it like this?

```
if not omit_opts:
    return __opts__.get(value, DEFAULTS.get(value, default))
```

Or perhaps

```
if not omit_opts:
    try:
        return __opts__[value]
    except KeyError:
        return DEFAULTS.get(value, default)
```

If you want to avoid always evaluating the `DEFAULTS.get(value, default)`.

Reply:
exception handling is slower than just checking if the value is in the list.

and .get is also slower.

This is the fastest implementation

Reply:
```
>>> import timeit
>>> setup = '__opts__ = {"one": 1, "two": 2}'
>>> timeit.timeit('''
... if 'one' in __opts__:
...     __opts__['one']
... ''', setup=setup)
0.047407952000000364
>>> timeit.timeit('''
... __opts__.get('one', 'THING')
... ''', setup=setup)
0.07143948800000999
```

Reply:
Huh! I wonder if that's something that's changed or if I've just been misremembering that `get` was faster than `in`, and it's just better than a `try/except`.

Well, TIL!

👍